drivers/CoreIMU/source/CoreLSM6DSOX.cpp

```diff
@@ -0,0 +1,110 @@
+// Leka - LekaOS
+// Copyright 2022 APF France handicap
+// SPDX-License-Identifier: Apache-2.0
+
+#include "CoreLSM6DSOX.h"
+
+namespace leka {
+
+CoreLSM6DSOX::CoreLSM6DSOX(mbed::I2C &i2c) : _i2c(i2c)
+{
+	_register_io_function.write_reg = (stmdev_write_ptr)ptr_io_write;
+	_register_io_function.read_reg	= (stmdev_read_ptr)ptr_io_read;
+	_register_io_function.handle	= static_cast<void *>(this);
+
+	init();
+}
+
+void CoreLSM6DSOX::init()
+{
+	lsm6dsox_init_set(&_register_io_function, LSM6DSOX_DRV_RDY);
+	lsm6dsox_i3c_disable_set(&_register_io_function, LSM6DSOX_I3C_DISABLE);
+	lsm6dsox_mode_get(&_register_io_function, nullptr, &_config);
+
+	_config.ui.xl.odr = _config.ui.xl.LSM6DSOX_XL_UI_26Hz_HP;
+	_config.ui.gy.odr = _config.ui.gy.LSM6DSOX_GY_UI_26Hz_HP;
+	_config.ui.xl.fs  = _config.ui.xl.LSM6DSOX_XL_UI_2g;
+	_config.ui.gy.fs  = _config.ui.gy.LSM6DSOX_GY_UI_2000dps;
+
+	lsm6dsox_mode_set(&_register_io_function, nullptr, &_config);
+
+	lsm6dsox_xl_data_rate_set(&_register_io_function, static_cast<lsm6dsox_odr_xl_t>(RateXL::_26Hz));
+	lsm6dsox_gy_data_rate_set(&_register_io_function, static_cast<lsm6dsox_odr_g_t>(RateGyr::_26Hz));
+	lsm6dsox_xl_full_scale_set(&_register_io_function, static_cast<lsm6dsox_fs_xl_t>(RangeXL::_2G));
+	lsm6dsox_gy_full_scale_set(&_register_io_function, static_cast<lsm6dsox_fs_g_t>(RangeGyr::_2000dps));
+}
+
+void CoreLSM6DSOX::setPowerMode(PowerMode mode)
+{
+	if (mode == PowerMode::Off) {
+		lsm6dsox_xl_data_rate_set(&_register_io_function, LSM6DSOX_XL_ODR_OFF);
+	} else {
+		lsm6dsox_xl_hm_mode_t power_mode {};
+		switch (mode) {
+			case PowerMode::UltraLow:
+				power_mode = LSM6DSOX_ULTRA_LOW_POWER_MD;
+				break;
+			case PowerMode::Normal:
+				power_mode = LSM6DSOX_LOW_NORMAL_POWER_MD;
+				break;
+			case PowerMode::High:
+				power_mode = LSM6DSOX_HIGH_PERFORMANCE_MD;
+				break;
+			default:
+				return;
+		}
+		lsm6dsox_xl_power_mode_set(&_register_io_function, power_mode);
+	}
+}
+
+auto CoreLSM6DSOX::updateData() -> lsm6dsox_data_t &
+{
+	lsm6dsox_data_get(&_register_io_function, nullptr, &_config, &data);
+	return data;
+}
+
+void CoreLSM6DSOX::turnOffSensors()
+{
+	lsm6dsox_embedded_sens_off(&_register_io_function);
+}
+
+auto CoreLSM6DSOX::read(uint8_t register_address, uint16_t number_bytes_to_read, uint8_t *p_buffer) -> int32_t
+{
+	// Send component address, with no STOP condition
+	auto ret = _i2c.write(_address, (const char *)&register_address, 1, true);
+	if (ret == 0) {
+		// Read data, with STOP condition
+		ret = _i2c.read(_address, (char *)p_buffer, number_bytes_to_read, false);
+	}
+
+	return ret;
+}
+
+auto CoreLSM6DSOX::write(uint8_t register_address, uint16_t number_bytes_to_write, uint8_t *p_buffer) -> int32_t
+{
+	std::array<uint8_t, 32> tmp {};
+
+	if (number_bytes_to_write >= 32) {
+		return 2;
+	}
```

Comment:
I2X write function returns 0 on success and non zero on failure

It can returns 1 (rather than magical number 2)  in order to turns it more "binary".